### ISSUE TITLE
Build fails on Windows with cygwin when some symbols are missing

### ISSUE THREAD

**Issue by epifanovsky** [flame/blis#252]
On Windows in the Cygwin environment, `configure auto` completes successfully, however the build subsequently fails at the library linking stage with errors that look like this:
```
$ make -j4
Dynamically linking lib/haswell/libblis.lib
/usr/lib/gcc/x86_64-pc-cygwin/7.3.0/../../../../x86_64-pc-cygwin/bin/ld: Cannot export bli_l3_thread_entry: symbol not defined
/usr/lib/gcc/x86_64-pc-cygwin/7.3.0/../../../../x86_64-pc-cygwin/bin/ld: Cannot export cblas_caxpy: symbol not defined
/usr/lib/gcc/x86_64-pc-cygwin/7.3.0/../../../../x86_64-pc-cygwin/bin/ld: Cannot export cblas_ccopy: symbol not defined
/usr/lib/gcc/x86_64-pc-cygwin/7.3.0/../../../../x86_64-pc-cygwin/bin/ld: Cannot export cblas_cdotc_sub: symbol not defined
...
```
This is because at the linking stage the linker is asked to export symbols from file `windows/build/libblis-symbols.def`, however it contains symbols that are not necessarily in the library. For example, the cblas symbols do not exist if BLIS is not configured with `--enable-cblas`.

@isuruf Any ideas?

**Comment by isuruf** [comment]
Yes, that's an issue with having just one def file. 
There are several options.

1. We can have multiple def files and combine them together depending on the configure options. (This is beneficial for all windows compilers.)
2. fix https://github.com/flame/blis/issues/248 so that .def files are not needed.

Short term option is to changing https://github.com/flame/blis/blob/d7537fb51dac0636591fc7c68261a2322642ab3c/common.mk#L502 to `-Wl,--export-all-symbols -Wl,--out-implib,$(LIBBLIS).dll.a` so that all symbols are exported and no .def is used.

**Comment by fgvanzee** [comment]
@isuruf Thanks for these comments and recommendations. I think the best long-term solution is for me to move forward with #248. (@epifanovsky and I were thinking of composing the final `.def` file together from components based on the `configure` options given, but, as you point out, that won't be necessary at all if we add configurable symbol export statements.)

Forgive this basic question, but I just want to make sure I understand: Is the purpose of the `.def` file for Windows builds to specify the complete list of symbols we wish to make public in the `.dll`, or the complete list of *all* symbols present in the `.dll`?

@epifanovsky I would say try out Isuru's short-term fix and then keep an eye on #248 for the long-term solution. Thanks for your feedback on this.

**Comment by isuruf** [comment]
>  the purpose of the .def file for Windows builds to specify the complete list of symbols we wish to make public in the .dll

It's this one. Any symbol that a user of blis or the testsuite use.
